### ISSUE TITLE
Content Security Policy failure caused by is-generator-function dependency

### ISSUE THREAD

**Issue by feross** [vercel/next.js#23587]
### What version of Next.js are you using?

10.1.2

### What version of Node.js are you using?

15.12.0

### What browser are you using?

Firefox

### What operating system are you using?

macOS

### How are you deploying your application?

Other

### Describe the Bug

"Content Security Policy: The page’s settings blocked the loading of a resource at eval (“script-src”)."

Caused by `is-generator-function`.

Output of `npm why is-generator-function`:

```
is-generator-function@1.0.8
node_modules/is-generator-function
  is-generator-function@"^1.0.7" from util@0.12.3
  node_modules/util
    util@"^0.12.0" from assert@2.0.0
    node_modules/assert
      assert@"2.0.0" from next@10.1.2
      node_modules/next
        next@"^10.1.2" from the root project
    util@"0.12.3" from next@10.1.2
    node_modules/next
      next@"^10.1.2" from the root project
```

### Expected Behavior

No CSP warnings.

### To Reproduce

When this code is run by next.js in Firefox, the warnings appears:

```js
	try {
		return Function('return function*() {}')();
	} catch (e) {
	}
```

It isn't clear what causes this to run, but you can see it in action in our app https://wormhole.app

**Comment by clement-faure** [comment]
Same here in Google Chrome (Version 100.0.4896.75 (Build officiel) (x86_64)) by following the https://github.com/vercel/next.js/tree/canary/examples/with-strict-csp exemple.

I'm getting this error : 

<img width="929" alt="Capture d’écran 2022-04-07 à 23 55 07" src="https://user-images.githubusercontent.com/23209488/162326302-e0e73a31-4bac-46d1-99ee-e8b356c48caa.png">

Versions : 
 `"next": "^12.1.4",`

**Comment by ljharb** [comment]
There's no way to avoid this, unfortunately, because generator function syntax has to be done with eval, or else it would hard-break older engines/browsers.

You could probably write a bundler transform, when you know that all the targets support the syntax natively, to replace that line in is-generator-function with the direct syntax.

**Comment by clement-faure** [comment]
Thank you for your answer. Unfortunately, i'm not able to reproduce the error message when cloning the https://github.com/vercel/next.js/tree/canary/examples/with-strict-csp, so i shouldn't have this message... I'll try to investigate on my side regarding my csp configuration...

**Comment by Kolombiken** [comment]
Should this issue maybe be renamed? This is still an issue but is not in any dependency as far as I can see but in the Next.js code. 

I'm also thinking there should be a [disclaimer about this in the documentation](https://nextjs.org/docs/advanced-features/security-headers#content-security-policy). At least as long as this is an issue. 
Right now the example there doesn't work at all and will give you warnings.

**Comment by ljharb** [comment]
I'd expect such a disclaimer in the _CSP_ documentation - disabling eval on the page is highly likely to break third-party code, and as such, should be carefully considered before restricting it.

**Comment by ljharb** [comment]
I totally agree, but not all eval is unsafe, unfortunately, which is the case here.

**Comment by Kolombiken** [comment]
I agree with you @ramosbugs. 
My point was that having an example in the documentation that will likely cause an error without any disclaimer will just cause confusion. And might result in people allowing more things than needed in their CSP making their application vulnerable. 

As far as I have tested it's possible to have pages without getting the error. But it's very likely that you will get an error. I'm not sure right now what the exact trigger is. Such information would also be helpful. E.g. if you are using feature x you need to enable `unsafe-eval` or your application will not work properly.

**Comment by StarpTech** [comment]
This is very bad-practice. `unsafe-eval` should never be required.  I run into this today as well. Can we prioritize it please. I'm on Next `14.1.0`.

**Comment by ljharb** [comment]
@StarpTech https://github.com/vercel/next.js/issues/23587#issuecomment-1309547758 eval using `Function` on a string literal is simply not unsafe.

**Comment by StarpTech** [comment]
But we can't whitelist this in the CSP policy. Using `unsafe-eval` will applied to all trusted origins.
